examples/model_selection/plot_grid_search_digits.py

```diff
@@ -13,56 +13,167 @@
-    print()
-
-    clf = GridSearchCV(SVC(), tuned_parameters, scoring="%s_macro" % score)
-    clf.fit(X_train, y_train)
-
-    print("Best parameters set found on development set:")
-    print()
-    print(clf.best_params_)
-    print()
-    print("Grid scores on development set:")
-    print()
-    means = clf.cv_results_["mean_test_score"]
-    stds = clf.cv_results_["std_test_score"]
-    for mean, std, params in zip(means, stds, clf.cv_results_["params"]):
-        print("%0.3f (+/-%0.03f) for %r" % (mean, std * 2, params))
-    print()
-
-    print("Detailed classification report:")
-    print()
-    print("The model is trained on the full development set.")
-    print("The scores are computed on the full evaluation set.")
-    print()
-    y_true, y_pred = y_test, clf.predict(X_test)
-    print(classification_report(y_true, y_pred))
-    print()
-
-# Note the problem is too easy: the hyperparameter plateau is too flat and the
-# output model is the same for precision and recall with ties in quality.
+scores = ["precision_macro", "recall_macro"]
+
+# %%
+# We can also define a function that will implement the strategy to use to
+# select the best candidate from the `cv_results_` returns from the
+# :class:`~sklearn.model_selection.GridSearchCV`. Once the candidate selected,
+# it will be automatically refitted by the
+# :class:`~sklearn.model_selection.GridSearchCV` instance. Here, the strategy
+# is to short list the model which are the best in terms of precision and
+# recall. From the selected models, we finally select the fastest model at
+# predicting.
+#
+# Note that these choices are completely arbitrary.
+import pandas as pd
+
+
+def refit_strategy(cv_results):
+    """Define the strategy to select the best estimator.
+
+    The strategy defined here is to filter-out all results below a precision threshold
+    of 0.98, rank the remaining by recall and keep all models with one standard
+    deviation of the best by recall. Once these models are selected, we can select the
+    fastest model to predict.
+
+    Parameters
+    ----------
+    cv_results : dict of numpy (masked) ndarrays
+        CV results as returned by the `GridSearchCV`.
+
+    Returns
+    -------
+    best_index : int
+        The index of the best estimator as it appears in `cv_results`.
+    """
+    # print the info about the grid-search for the different scores
+    precision_threshold = 0.98
+    for score in scores:
+        print(f"\nGrid scores on development set using {score}:\n")
+        for mean, std, params in zip(
+            cv_results[f"mean_test_{score}"],
+            cv_results[f"std_test_{score}"],
+            cv_results["params"],
+        ):
+            print(f"{mean:0.3f} (+/-{std * 2:0.03f}) for {params}")
+
+    cv_results_ = pd.DataFrame(cv_results)
+
+    # Filter-out all results below the threshold
+    cv_results_ = cv_results_[
+        cv_results_["mean_test_precision_macro"] > precision_threshold
+    ]
```

Comment:
I think we should print the filtered list of high precision models here (with the value of the precision threshold).

```python
print(f"Models with a precision higher than {precision_threshold}:")
print_dataframe(high_precision_cv_results)
```